### PR TITLE
harness: support mocking storage mounts

### PR DESCRIPTION
**Pull request by rwcarlsen**
This mocks storage mounts (e.g. storage from metadata.yaml) with using a disk-backed tmpdir.  It includes support for pebble file-ops bridging between workload and charm container - i.e. pebble file ops can "see" files in a workload storage mount that were put there directly via a (shared) storage mount in a charm container.

### CODE REVIEW

File: docs/storage.md
Comment (pengale) [inline docs/storage.md:62]: It feels like this paragraph needs another pass for simplicity and clarity.

For example, this is a tangled sentence snippet:

> Storage with this type of `multiple:...` configuration will have each instance residing under ..."

File: ops/testing.py
Comment (sed-i) [inline ops/testing.py:1245]: Should we protect here against `cd`-ing out of `/tmp`?
```python
>>> os.path.join('/tmp/temp', '//weird//name')
'//weird//name'
```

File: ops/testing.py
Comment (sed-i) [inline ops/testing.py:1079]: Does this require an explicit `cleanup()`? To be added to the harness cleanup method?

File: docs/storage.md
Comment (rwcarlsen) [reply docs/storage.md:62]: Not going to disagree with you here - this was just a first draft brain dump that I'll eventually get into discourse.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1245]: Perhaps we could add an easter egg test writers: `error: why you puttin' slashes in test container names?`

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1079]: No - it's auto cleaned up when the object dies - or when the context ends if it's used inside one.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1271]: I added a check that errors for absolute paths

File: docs/storage.md
Comment (rwcarlsen) [reply docs/storage.md]: I've tried to clean up this doc content and put it up on discourse here: https://discourse.charmhub.io/t/storage/4466.

File: ops/model.py
Comment (jnsgruk) [inline ops/model.py:1887]: Maybe I'm being dense, but this argument doesn't seem to actually get used anywhere?

File: ops/testing.py
Comment (jnsgruk) [inline ops/testing.py:1113]: How distinct is this from `self._pebble_clients` - these seem very closely named and the type annotations seem to suggest a pretty similar structure, too!

File: ops/testing.py
Comment (jnsgruk) [inline ops/testing.py:1272]: Nit: we already had `pathlib` imported in this file, and iirc it's used fairly extensively elsewhere in the codebase to join paths together?

```suggestion
                "location": pathlib.Path(self._harness_tmp_dir.name) / name / str(index),
```

File: ops/testing.py
Comment (jnsgruk) [reply ops/testing.py:1113]: Reading a bit further down about how this is used, you could almost certainly derive the right value from the socket path that's used as the key in the existing dict? I don't know if that's actually better, just feels like we're duplicating here.

File: ops/model.py
Comment (jnsgruk) [reply ops/model.py:1887]: Ah, looks like this is in place to keep the interface the same across real pebble and harness pebble?

Comment (rwcarlsen) [reply ops/model.py:1887]: Yep

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1113]: I could easily be convinced either way.  Relevant considerations:

* At least one other place in the code uses the pebble_clients dict by rebuilding a socket path starting from a container name.  This makes it seem like a container_name->client mapping might be more natural.
* our socket path conventions could change (but unlikely)
* non-standard socket paths get used in testing.

The last point could be easily remedied with test mods.  If we believe the first point, then it would be worth "fixing" this spot: https://github.com/canonical/operator/blob/e3bc30bf8491716d33c8bb6061eb63bd8306ca1d/ops/testing.py#L759

Comment (jnsgruk) [reply ops/testing.py:1113]: Okay so point 1 makes me wonder whether there is a drive-by refactor in order here? I'm not hard over either way, this shouldn't block merging.

**Review by pengale — APPROVED** [review]
There's already been some design and philosophy discussion, so I did a pass just to see if any major questions or potential bugs jumped out at me.

Nothing did. I'm +1! 

(I like Jon's suggestion to use pathlib, but I am +1 w/ or without it.)

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1274]: There's other code that would need to be changed to accommodate this; also only changing this attrib would leave logic around this more bifurcated on path handling.  I think this is probably better left for another more filepath focused PR.

Comment (rwcarlsen) [reply ops/testing.py:1274]: https://github.com/canonical/operator/issues/742

Comment (jnsgruk) [reply ops/testing.py:1274]: Ack, thanks.

File: ops/model.py
Comment (jnsgruk) [inline ops/model.py:1042]:
```suggestion
        id: The index number for storage
```

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1347]: the problem with having both is that it is entirely plausible to now have them disagree. (eg one thing called 'get_pebble(name, path)' and another did 'get_pebble(name, other_path)'. (yes, someone doing that is doing it wrong, but it *is* the api we are now exposing.
)

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1792]: I would rather not call this _Mock, as it doesn't act like a Mock (record, replay, check the calls, etc)
We specifically named it `_TestingModelBackend`, mabye `_TestingStorage` ? (Sorry I missed this for `_MockFilesystem`, maybe we should rename that one while we're here? Since it also isn't really act like a Mock). At the very least Fake would be  better than Mock for terminology here.


I would definitely like to see a docstring on various bits. Like what is 'location' vs what is 'src', etc.
I believe I know the answer today, but it won't be obvious in another month. :)

Also defining that `src` will be created if it doesn't exist is a pretty important point.

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1799]: This could also use a docstring. Because it defines a contains as in 'exists inside my directory'. But it doesn't mean that the path actually exists.
Also, I *think* there is the potential for it being wrong as:
```
'/foo/bar/baz'.startswith('/foo/b')
```
*If* we require that _location ends with a '/' then we should document it, and enforce it in `__init__`. (And, in fact, I would guess the default is just the location defined in the charm metadata, and 99% of the time does *not* end in a '/'.)

We could also just use `pathlib` and use things like:  `is_relative_to(other)`
https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to

IIRC pathlib is in python 3.5, but not all OS and System functions allow you to use it. (which is why a bunch of stuff uses `os.SOMETHING(str(path))`. But I would still recommend using pathlib for the Storage object.

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1828]: This is a dangerous 
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement

Since these are disabled sometimes, I think we'd rather just have an 
```
if not self.contains(path): raise RuntimeError(...)
```
It also gives you way to give a helpful error message instead of just an assertion failure. (what is the `location` what is the `path` that was supplied, etc.)

Comment (jameinel) [reply ops/testing.py:1828]:
```
>>> def contains(foo):
...   foo.startswith('/bar')
...
>>> assert(contains('/ba'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```
Definitely doesn't give anyone a way to debug what they did wrong.

Since many call sites are doing this, I think a helper self._ensure_path_is_contained() is reasonable.

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1852]: I believe you can do `list_dir` for subdirectories as well. eg:
```
location = '/foo'
storage.mkdir('bar')
storage.mkdir('bar/baz')
storage.list_dir('bar/baz')
```

I don't see much in the way of direct testing of _MockStorage. This is probably something that should be covered to make sure we work correctly. (My second look at it says that it might, but I thought it would probably be good to verify)

I also think this code gets slightly cleaner with `pathlib`, so would recommend it.
(Note that locations can be used with pathlib.PurePosixPath, but src probably needs to use pathlib.Path)

File: ops/testing.py
Comment (jameinel) [inline ops/testing.py:1854]: _Directory actually defines its parameter as PurePosixPath not a plain str.
in current python
```
>>> import pathlib
>>> etc = pathlib.PurePosixPath('/etc')
>>> import os
>>> os.path.join(etc, 'foo')
'/etc/foo'
```
So os.path will accept pathlib, but they do always return a str.

File: test/test_testing.py
Comment (jameinel) [inline test/test_testing.py:146]: you should really have a self.addCleanup here rather than just setting it at the end.
because if the test fails, then the value is in an indeterminate state.

**Review by jameinel — COMMENTED** [review]
I think there are a few things that are relatively minor, and some that I feel a bit more strongly about.
I think this is definitely a reasonable first step for testing Storage interaction. I like that it does share the mount points as would happen in the real filesystem.

File: ops/model.py
Comment (jnsgruk) [reply ops/model.py:1903]: Yeh, I just realised I failed to come back on this. This *does* feel a little uncomfortable to me - and it'll certainly appear odd to others reading the codebase that there is a random unused argument here (as it did to me until I read the whole PR..)

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:1115]: This is just food for thought and a general thing we might consider to start doing everywhere:
I recently heard a rather convincing talk by one of the core python devs who recommended doing the following:

Instead of:
```python
 # {<container_name>: <client>}
 self._pebble_containers = {}  # type: {str: _TestingPebbleClient}
```
we could do:

```python
# somewhere on top of the file:
ContainerName = str
# ...

self._pebble_containers: Dict[ContainerName: _TestingPebbleClient] = {}
```

I quite like it.
About this specific line I +1 what's already been said above.

File: ops/testing.py
Comment (PietroPasotti) [inline ops/testing.py:1265]: is it **test storage** specifically, or storage names in general that cannot start with /? If the former: why the discrepancy? If the latter: remove 'test'.

File: ops/testing.py
Comment (PietroPasotti) [inline ops/testing.py:1274]: If `_harness_tmp_dir: Path`: 
`self._harness_tmp_dir / name / str(index)`

File: ops/testing.py
Comment (PietroPasotti) [inline ops/testing.py:1295]: again the name `storage_id` is no longer accurate, we should be uniform in how we call that thing, whatever it is

File: ops/testing.py
Comment (PietroPasotti) [inline ops/testing.py:1824]: I'd expect anything marked as optional to have None as default,
I'd do:

- `encoding: str = None`
- specify in the docstring that the default is 'utf-8'
- later do: `encoding = encoding or 'utf-8'`

However it seems to me that you really do need an encoding one way or another, so why not just
`encoding: str = 'utf-8'`

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:1854]: I like `etc / 'foo'` as well

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
Overall looks good, I only have some concerns about the consistency of the naming of what used to be `storage_id` in method names, var names and docstrings, and a couple of points where a bug might be hidden or a more helpful error message should be raised.
I think we need another quick iteration on these points.

File: ops/model.py
Comment (rwcarlsen) [reply ops/model.py]: I've reverted this (private) api change.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: I consolidated the dictionaries to one with a container name index and updated all relevant code.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: This was reverted as noted above.  The one place in the test harness where we need the container name, I just parse it out of the socket path.  Not the greatest, but should work well enough.  We can always update that parsing if the socket path changes.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: Indeed you are right on the potential bug.  Should be addressed now - I've updated the entire _TestingStorage class to use pathlib everywhere.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: Pathlib conversion done.  I'll make sure tests get well rounded out too.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1871]: Done.  All converted to pathlib.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:212]: I pushed up a tweak to this example that should make it a bit clearer.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1261]: I actually think you'd probably have problems if you tried to put a slash anywhere in the storage name since they get munged with paths a lot.  I don't know if this is a hard rule for juju though.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: harness_tmp_dir is a tempfile.TemporaryDirectory.  It's not a string or a pathlib path.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1291]: It actually is accurate.  I am trying to make things consistent for users. There is other api and docs that uses "storage_id" to refer to the fully-qualified "<name>/<index>" form.  So the most consistent/cleanest way I saw to improve things for users was to preserve this terminology while trying to make the framework internals change to use the name+index convention.  So id=whole-thing.  name=name.  index=number.  Basically I made it so that all the places (including user facing ones) that were already using storage_id to refer to the entire label (including index) didn't have to change.  The only user facing change is deprecating the super confusing, overloaded "id" attribute on the storage class.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1841]: Yeah - this was a copy-paste holdover.  Fixed.

File: ops/model.py
Comment (rwcarlsen) [reply ops/model.py:1047]: The issue here is that "storage_id" from juju and within the framework *is* the fully qualified "<name>/<number>" identifier.  All the places in the framework that I've changed to be "storage_index" were actually only referring to the integer part of the full id.  So the rename was just to clear up this overloading/confusion.

Comment (rwcarlsen) [reply ops/model.py:1047]: That said - I'm definitely up for alternative names instead of "storage_index".  But it can't be "id", because that is already taken.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:476]: No.  storage_indices comes from calling storage_add - which only returns the indices for the newly-added storage.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1262]: In fact, this code https://github.com/canonical/operator/blob/e3bc30bf8491716d33c8bb6061eb63bd8306ca1d/ops/main.py#L154-L157 would clearly be broken if the storage name contained a `/`.  This is worth looking into further...

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py]: Alright - I've converted the testingfilesystem tests to a mixin that is now applied to both the TestingFilesystem and the TestingStorageMount classes (and fixed a few inconsistencies the tests caught).

File: ops/charm.py
Comment (rwcarlsen) [reply ops/charm.py:497]: Added relevant error checking.

**Comment by jnsgruk** [comment]
Failing test and query on '/' in storage names aside, this now looks much cleaner. Thanks for the iteration!

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:1262]: Okay.  Found the answer.  Juju uses this to validate storage names: https://github.com/juju/names/blob/9c6532a528238568c78ebb8a41bbb20340da3f90/storage.go#L17-L20 so no slashes allowed.

Comment (rwcarlsen) [reply ops/testing.py]: I updated the error check to just disallow slashes in anywhere in the name.

**Review by jnsgruk — APPROVED** [review]

File: ops/model.py
Comment (PietroPasotti) [reply ops/model.py:1063]: `full_name`?  and I know `identifier` is long for `id`, but if the long-term plan is to replace what used to be 'id' with what now is 'full_id' (so I thought), we could do:

old version:
id: X

current version:
id: X (deprecated: use self.index)
identifier: Y 
index: X

future version:
identifier: Y (deprecated: use id)
id: Y
index: X

future future version:
id: Y
index: X


Otherwise we go immediately for a hard replace of the name `id`, and wait for a major version bump to put this in.

Comment (PietroPasotti) [reply ops/model.py:1063]: either way this is a non-blocking nit, feel free to ignore it

**Review by PietroPasotti — APPROVED** [review]
LGTM! Pending that name nit, but don't consider it blocking

File: ops/model.py
Comment (rwcarlsen) [reply ops/model.py:1063]: I think I'm going to stick with `full_id`.  It's (somewhat) descriptive and retains the `id` suffix which helps the connection between this and `storage_id` be more obvious to users of our apis.

Comment (jnsgruk) [reply ops/model.py:1063]: +1

File: test/test_testing.py
Comment (rwcarlsen) [reply test/test_testing.py:151]: done

**Pull request merged**
